datalad/core/dataset/gitrepo.py

```diff
@@ -0,0 +1,519 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Internal low-level interface to Git repositories
+
+"""
+
+import logging
+from os import environ
+import re
+import time
+from weakref import WeakValueDictionary
+
+from datalad.config import (
+    ConfigManager,
+)
+from datalad.core.dataset import (
+    PathBasedFlyweight,
+    RepoInterface,
+    path_based_str_repr,
+)
+from datalad.utils import (
+    ensure_list,
+    Path,
+)
+from datalad.support.exceptions import (
+    CommandError,
+    GitIgnoreError,
+    InvalidGitRepositoryError,
+)
+from datalad.cmd import (
+    GitRunner,
+    run_gitcommand_on_file_list_chunks,
+)
+
+lgr = logging.getLogger('datalad.core.dataset.gitrepo')
+
+
+@path_based_str_repr
+class GitRepo(RepoInterface, metaclass=PathBasedFlyweight):
+    """Representation of a git repository
+
+    """
+    # We must check git config to have name and email set, but
+    # should do it once
+    _config_checked = False
+
+    # Begin Flyweight:
+
+    _unique_instances = WeakValueDictionary()
+
+    GIT_MIN_VERSION = "2.19.1"
+    git_version = None
+
+    def _flyweight_invalid(self):
+        return not self.is_valid_git()
+
+    @classmethod
+    def _flyweight_reject(cls, id_, *args, **kwargs):
+        # TODO:
+        # This is a temporary approach. See PR # ...
+        # create = kwargs.pop('create', None)
+        # kwargs.pop('path', None)
+        # if create and kwargs:
+        #     # we have `create` plus options other than `path`
+        #     return "Call to {0}() with args {1} and kwargs {2} conflicts " \
+        #            "with existing instance {3}." \
+        #            "This is likely to be caused by inconsistent logic in " \
+        #            "your code." \
+        #            "".format(cls, args, kwargs, cls._unique_instances[id_])
+        pass
+
+    # End Flyweight
+
+    def __init__(self, path):
+        """Creates representation of git repository at `path`.
+
+        Parameters
+        ----------
+        path: str
+          path to the git repository; In case it's not an absolute path,
+          it's relative to PWD
+        """
+        self.path = path
```

Comment:
See above. `path` related stuff would need to involve all three classes, of course.